### PR TITLE
Add threadsafe Pool data structure

### PR DESCRIPTION
**Pull request by quinnj**
Also split the now-large ConcurrentUtilities.jl file into more manageable individual files. For reviewing, mainly worry about the `src/pools.jl` and `test/pools.jl` file, since they're the only net-new code here.

### CODE REVIEW

File: src/pools.jl [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    # if we're given an object, we'll put it back in the pool
+    # otherwise, we'll just return the permit
+    if obj !== nothing
+        # first though, we repeat Base.Semaphore's error check in the case of an invalid release
+        # where we don't want to push an object for reuse for an invalid release
+        Base.@lock pool.sem.cond_wait begin
+            if pool.sem.curr_cnt > 0
+                # if the key is invalid, we'll just let the KeyError propagate
+                objs = Base.@lock pool.values pool.values[][key]
+                push!(objs, obj)
```
Comment (Drvi) [inline src/pools.jl:96]: So the `push!` method allocates a new `Node` object every time (see the benchmarks shared above) which might be necessary for things to be threadsafe (?), but when `max` is provided by the user, IIUC, there is an upper bound of `Node`s that can be populated at any given time, so in that case we could reuse the `Node`s we allocated before? Like when we `pop!` a `Node`, we'd just keep it around and reuse it later?

**Review by Drvi — COMMENTED** [review]
Left a couple of comments about optimization opportunities. Would it make sense to specialize the Pool on being keyed/non-keyed and Semaphore-bounded vs not?

File: src/pools.jl [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    sem::Semaphore
+    values::Lockable{Dict{K, ConcurrentStack{T}}}
+end
+
+Pool(T; max::Int=typemax(Int)) = Pool{Nothing, T}(Semaphore(max), Lockable(Dict{Nothing, ConcurrentStack{T}}()))
+Pool(K, T; max::Int=typemax(Int)) = Pool{K, T}(Semaphore(max), Lockable(Dict{K, ConcurrentStack{T}}()))
+
+Base.empty!(pool::Pool) = Base.@lock pool.values empty!(pool.values[])
+
+TRUE(x) = true
```
Comment (nickrobinson251) [inline src/pools.jl:37]: optional:
since this is yet another thing we can change when we drop v1.6, I sometimes like to leave comments so i can grep for "VERSION" when eventually dropping old Julia versions

```suggestion
# in VERSION >= v"1.7", we can replace `TRUE` with `Returns(true)`
TRUE(x) = true
```

File: test/pools.jl [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    yield()
+    @test !istaskdone(tsk)
+    # release an object back to the pool
+    release(pool, x1)
+    # now the acquire can complete
+    x1 = fetch(tsk)
+    # even though we released 1 for reuse, we passed forcenew, so our function was called to create new
+    @test x1 == 4
+    # error to try and provide a key to a non-keyed pool
+    @test_throws ArgumentError acquire(() -> 1, pool, 1)
```
Comment (nickrobinson251) [inline test/pools.jl:31]: I tried stepping through these tests locally, and i found the amount of state quite inconvenient

it made it hard to follow which tests depended on what came before and which didn't

can we introduce `@testset`s to make these things clearer, and create a fresh `Pool` for each testset e.g. this could just be
```suggestion
    @testset "keyed acquire on non-keyed pool" begin
        # error to try and provide a key to a non-keyed pool
        pool = Pool(Int, max=2)
        @test_throws ArgumentError acquire(() -> 1, pool, 1)
    end
```
```

File: test/pools.jl [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    release(pool, x1)
+    release(pool, x2)
+    release(pool, x3)
+    # try to do an invalid release
+    @test_throws ErrorException release(pool, 10)
+    # test that the invalid release didn't push the object to our pool for reuse
+    x1 = acquire(() -> 8, pool)
+    @test x1 == 7
+    # calling empty! removes all objects for reuse
+    empty!(pool)
```
Comment (nickrobinson251) [inline test/pools.jl:58]: why do we support `empty!` ? 

At first i thought `Pool` was supporting the exact same interface as `Semaphone`, which is just `acquire`/`release`... but then there's `empty!`

i think i'm confused by what the contract of `empty!` is supposed to be here (see my comment above). 
what does "removes all objects _for reuse_" mean? I wonder if `empty!` is even the right name.

It also makes me wonder if there should also be a `release_all(pool)` that just releases all the permits?

And if there should be other Dict-like interface functions (e.g. `keys`/`haskey` or `length`)?

File: src/pools.jl [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+"""
+    acquire(f, pool::Pool{K, T}, [key::K]; forcenew::Bool=false, isvalid::Function) -> T
+
+Get an object from a `pool`, optionally keyed by the provided `key`. If no pool exists for the given key, one will be created.
```
Comment (nickrobinson251) [inline src/pools.jl:42]:
```suggestion
Get an object from a `pool`, optionally keyed by the provided `key`. If no object exists for the given key, one will be created.
```

File: src/pools.jl [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+The provided function `f` must create a new object instance of type `T`.
+The acquired object MUST be returned to the pool by calling `release(pool, key, obj)` exactly once.
```
Comment (nickrobinson251) [inline src/pools.jl:44]: Do we need to call 3-arg `release`? isn't it just:
```suggestion
Each `acquire` call MUST be matched by exactly one `release` call.
```

File: src/pools.jl [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+end
+
+"""
+    release(pool::Pool{K, T}, key::K, obj::Union{T, Nothing}=nothing)
+    release(pool::Pool{K, T}, obj::T)
+    release(pool::Pool{K, T})
+
+Return an object to a `pool`, optionally keyed by the provided `key`.
+If `obj` is provided, it will be returned to the pool for reuse. Otherwise, if `nothing` is returned,
+just the "permit" will be returned to the pool.
```
Comment (nickrobinson251) [inline src/pools.jl:83]: so for a keyed pool to return just the permit you have to call `release(pool, key, nothing)` but for a non-keyed pool you have to call `release(pool)`? that seems inconsistent to me

File: src/lockable.jl [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+and release the lock when `f` returns. `f` will receive one positional
+argument: the value wrapped by `l`. If the lock is already locked by a
+different task/thread, wait for it to become available.
+When this function returns, the `lock` has been released, so the caller should
+not attempt to `unlock` it.
+"""
+function Base.lock(f, l::Lockable)
+    lock(l.lock) do
+        f(l.value)
+    end
```
[the remark this answers is not in the document]
Comment (quinnj) [reply src/lockable.jl:29]: Yeah, let's do it!

File: src/pools.jl [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+By default, `release(pool, obj)` will return the object to the pool for reuse.
+`release(pool)` will return the "permit" to the pool while not returning
+any object for reuse.
+"""
+struct Pool{K, T}
+    sem::Semaphore
+    values::Lockable{Dict{K, ConcurrentStack{T}}}
+end
+
+Pool(T; max::Int=typemax(Int)) = Pool{Nothing, T}(Semaphore(max), Lockable(Dict{Nothing, ConcurrentStack{T}}()))
```
[the remark this answers is not in the document]
Comment (quinnj) [reply src/pools.jl:32]: We might be able to do something faster in the non-key case, which I think is worth exploring (like having an extra field for a single stack that we use if the key type is `Nothing`)

File: src/pools.jl (hunk 1 above)
Comment (quinnj) [reply src/pools.jl:96]: Hmmmmm......maybe? We'd have to make the `Node`s mutable so you could reset the values. Maybe we could pass the `max` along to the `ConcurrentStack` struct and it could create its own little pool of `Node`s that would get reused? Probably worth exploring to see if we can avoid the allocations.

File: src/pools.jl (hunk 7 above)
Comment (quinnj) [reply src/pools.jl:83]: I can update the comment, but you can call `release(pool, key)` on a keyed pool.

File: src/pools.jl (hunk 1 above)
Comment (Drvi) [reply src/pools.jl:96]: Yeah sounds good! I think getting rid of allocs could have a nice impact on performance as well. Nodes are already mutable so its really about managing the pool... which is becoming a theme here:)

File: src/pools.jl (hunk 5 above)
Comment (quinnj) [reply src/pools.jl:42]: I think I'll just take this line out since it's confusing; we're not creating an pool or object, but a ConcurrentStack which we'll use to store the object once it comes back.

File: test/pools.jl (hunk 4 above)
Comment (quinnj) [reply test/pools.jl:58]: Yeah, let's come up with a better name; HTTP has this function `closeall`, but that's not quite the right thing I think for a generic pool. Maybe `evictall`? `empty!` felt the most right to me, but as you pointed out there's this tension between whether a `Pool` is just a semaphore or a Dict storing non-checked-out objects.

File: src/pools.jl (hunk 1 above)
Comment (Drvi) [reply src/pools.jl:97]: It would be nice to get rid of the allocs even when `max` is not provided so maybe a small pool could be maintained even in that case?

File: src/lockable.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (quinnj) [reply src/lockable.jl:29]: On my updated PR (refactored pool) using the above benchmark, I'm seeing:
```julia
julia> @btime foo()
  673.667 μs (8 allocations: 132.39 KiB)
```

File: src/pools.jl (hunk 9 above)
[the remark this answers is not in the document]
Comment (quinnj) [reply src/pools.jl]: Updated timing post-refactor of the above benchmark:
```julia
julia> @btime foo()
  3.740 ms (8 allocations: 132.39 KiB)
Pool{Nothing, Int64}(Base.GenericCondition{ReentrantLock}(Base.IntrusiveLinkedList{Task}(nothing, nothing), ReentrantLock(nothing, 0x00000000, 0x00, Base.GenericCondition{Base.Threads.SpinLock}(Base.IntrusiveLinkedList{Task}(nothing, nothing), Base.Threads.SpinLock(0)), (12800, 0, 141733920768))), 10000, 0, #undef, [10001, 10002, 10003, 10004, 10005, 10006, 10007, 10008, 10009, 10010  …  19991, 19992, 19993, 19994, 19995, 19996, 19997, 19998, 19999, 20000])

```

File: src/pools.jl (hunk 1 above)
Comment (quinnj) [reply src/pools.jl]: There are now no allocations in `acquire`/`release` on my refactored PR

File: test/pools.jl (hunk 4 above)
Comment (quinnj) [reply test/pools.jl]: Changed the function name to `drain!` in the PR

File: src/lockable.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (Drvi) [reply src/lockable.jl:29]: Very cool:) Still, I think the macro is preferable

File: src/pools.jl (hunk 9 above)
[the remark this answers is not in the document]
Comment (Drvi) [reply src/pools.jl]: Love it! This is much more lean:)
I hope this is as performant in your HTTP.jl benchmarks -- the dummy benchmarks I used were not stressing any concurrent accesses. I think we're now spending more time behind lock, which originally caused you to look into lockfree datastructures.

[the remark this answers is not in the document]
Comment (quinnj) [reply src/pools.jl]: That was mainly because we were calling `connect` while holding the lock, which then depends on network conditions and can easily take milliseconds. I think we should be ok here. The only slight worry is that we do call `isvalid` while holding the lock, so users have to be careful that their it isn't expensive to call that (it should be fine for how we use it in HTTP)

File: src/pools.jl [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+The `isvalid` keyword argument can be used to specify a function that will be called to determine if an object is still valid
+for reuse. By default, all objects are considered valid.
+If there are no objects available for reuse, `f` will be called to create a new object.
+If the pool is already at its maximum capacity, `acquire` will block until an object is returned to the pool via `release`.
+"""
+function Base.acquire(f, pool::Pool{K, T}, key=nothing; forcenew::Bool=false, isvalid::Function=TRUE) where {K, T}
+    key isa K || keyerror(key, K)
+    Base.@lock pool.lock begin
+        # first get a permit
+        while pool.cur >= pool.max
```
Comment (Drvi) [inline src/pools.jl:101]: Let's assume noone holds the lock, but task (1) is waiting on the `pool.lock` condition. Is it racy when task (2) releases from the pool, just as task (3) also tries to acquire? Task 1 would wake up, task 3 would hold the lock and both of them might now pass the  `pool.cur >= pool.max` check?

Comment (quinnj) [reply src/pools.jl:101]: > Task 1 would wake up, task 3 would hold the lock

This isn't possible. The design of `Threads.Condition` means that when you "wake up", you are guaranteed exclusive access to the lock.
